src/transformers/models/molmo/modular_molmo.py

```diff
@@ -0,0 +1,2289 @@
+        vision_feature_layers=(-2, -9),
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.image_token_index = image_token_index
+        self.image_seq_length = image_seq_length
+        self.vision_feature_select_strategy = vision_feature_select_strategy
+        self.vision_feature_layers = vision_feature_layers
+        if vision_config is None:
+            vision_config = {}
+            logger.info("vision_config is None. initializing the MolmoVisionConfig with default values.")
+        if text_config is None:
+            text_config = {}
+            logger.info("text_config is None. initializing the MolmoTextConfig with default values.")
+        if pooling_config is None:
+            pooling_config = {}
+            logger.info("pooling_config is None. initializing the MolmoPoolingConfig with default values.")
+        self.vision_config = MolmoVisionConfig(**vision_config)
+        self.text_config = MolmoTextConfig(**text_config)
+        self.pooling_config = MolmoPoolingConfig(**pooling_config)
+        self.initializer_range = initializer_range
+
+    @classmethod
+    def from_text_vision_configs(
+        cls,
+        text_config: MolmoTextConfig,
+        vision_config: MolmoVisionConfig,
+        pooling_config: MolmoPoolingConfig,
+        **kwargs,
+    ):
+        r"""
+        Instantiate a [`MolmoConfig`] (or a derived class) from molmo text model configuration, molmo vision model
+        configuration and molmo pooling module conffiguration.
+
+        Returns:
+            [`MolmoConfig`]: An instance of a configuration object
+        """
+
+        return cls(
+            text_config=text_config.to_dict(),
+            vision_config=vision_config.to_dict(),
+            pooling_config=pooling_config.to_dict(),
+            **kwargs,
+        )
+
+
+class MolmoCausalLMOutputWithPast(LlavaCausalLMOutputWithPast):
+    pass
+
+
+# swiglu activation
+class MolmoSwiGLU(nn.Module):
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x, gate = x.chunk(2, dim=-1)
+        return nn.functional.silu(gate) * x
+
+
+# text modules inherited from Qwen2
+class MolmoMLP(CLIPMLP):
+    def __init__(self, config):
+        super().__init__()
+        self.activation_fn = MolmoSwiGLU()
+        self.fc1 = nn.Linear(config.hidden_size, config.intermediate_size, bias=False)
+        self.fc2 = nn.Linear(config.intermediate_size // 2, config.hidden_size, bias=False)
+
+
+# We have different attention classes for the txt and the image components, they need to be propagated back correctly
+class MolmoTextAttention(Qwen2Attention):
+    pass
+
+
+class MolmoTextSdpaAttention(MolmoTextAttention, Qwen2SdpaAttention):
+    pass
+
+
+class MolmoTextFlashAttention2(MolmoTextAttention, Qwen2FlashAttention2):
+    pass
+
+
+MOLMO_TEXT_ATTENTION_CLASSES = {
+    "eager": MolmoTextAttention,
+    "sdpa": MolmoTextSdpaAttention,
+    "flash_attention_2": MolmoTextFlashAttention2,
+}
+
```

Comment:
```suggestion
```
you don't need them

Reply:
we needed them to get MolmoTextAttention instead of MolmoAttention to indicate the attn module is for text only. These are new changes from @Cyrilvallez to make modular happy when the object name has special Text/Vision suffixes